### PR TITLE
[dtensor] support aten.where and enable implicit scalar promotion

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110585
* __->__ #110584

This PR adds support for aten.where and support implicit scalar
promotion, basically when we meet scalar tensors in dispatching logic,
we implicitly convert it those to replicated dtensor

The latter also enables bunch of ops in op db to pass

### CODE REVIEW

File: torch/testing/_internal/distributed/_tensor/common_dtensor.py [hunk 1]
```diff
@@ -336,21 +336,11 @@ def to_dist_tensor(
...
-                        mesh,
-                        tuple(placements),
-                        size=t.size(),
-                        dtype=torch.bool,
-                        requires_grad=t.requires_grad,
-                        stride=t.stride()
-                    )
+                if t.ndim == 0:
+                    # scalar tensor by default will be replicated
+                    r = distribute_tensor(t, mesh, [Replicate()] * mesh.ndim)
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/testing/_internal/distributed/_tensor/common_dtensor.py:341]: bool type tensor was not supported by gloo backend that's why the special handling before, @kumpera added gloo bool support a while ago so now the special handling not needed

**Comment by wanchaol** [comment]
@fduwjj could you stamp the PR if there's no other comments :)

File: torch/testing/_internal/distributed/_tensor/common_dtensor.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/testing/_internal/distributed/_tensor/common_dtensor.py:341]: Ah I see. Make sense.

**Review by fduwjj — APPROVED** [review]
LGTM

**Pull request closed without merging**